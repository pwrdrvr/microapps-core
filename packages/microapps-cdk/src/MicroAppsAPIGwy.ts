import * as apigwy from '@aws-cdk/aws-apigatewayv2';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as iam from '@aws-cdk/aws-iam';
import * as logs from '@aws-cdk/aws-logs';
import * as r53 from '@aws-cdk/aws-route53';
import * as r53targets from '@aws-cdk/aws-route53-targets';
import * as cdk from '@aws-cdk/core';

export interface MicroAppsAPIGwyProps {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: cdk.RemovalPolicy;

  /**
   * CloudFront edge domain name
   *
   * @example apps.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameEdge?: string;

  /**
   * API Gateway origin domain name
   *
   * @example apps-origin.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameOrigin?: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Optional local region ACM certificate to use for API Gateway
   * Note: required when using a custom domain
   *
   * @default none
   */
  readonly certOrigin?: acm.ICertificate;

  /**
   * Route53 zone in which to create optional `domainNameEdge` record
   */
  readonly r53Zone?: r53.IHostedZone;

  /**
   * Path prefix on the root of the API Gateway Stage
   *
   * @example dev/
   * @default none
   */
  readonly rootPathPrefix?: string;
}

export interface IMicroAppsAPIGwy {
  /**
   * Domain Name applied to API Gateway origin
   */
  readonly dnAppsOrigin?: apigwy.IDomainName;

  /**
   * API Gateway
   */
  readonly httpApi: apigwy.HttpApi;
}

export class MicroAppsAPIGwy extends cdk.Construct implements IMicroAppsAPIGwy {
  private _dnAppsOrigin: apigwy.DomainName | undefined;
  public get dnAppsOrigin(): apigwy.IDomainName | undefined {
    return this._dnAppsOrigin;
  }

  private _httpApi: apigwy.HttpApi;
  public get httpApi(): apigwy.HttpApi {
    return this._httpApi;
  }

  /**
   * MicroApps - Create just API Gateway
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: cdk.Construct, id: string, props?: MicroAppsAPIGwyProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props cannot be undefined');
    }

    if (
      (props.r53Zone === undefined && props.domainNameEdge !== undefined) ||
      (props.r53Zone !== undefined && props.domainNameEdge === undefined)
    ) {
      throw new Error('If either of r53Zone or domainNameEdge are set then the other must be set');
    }

    if (
      (props.domainNameOrigin === undefined && props.certOrigin !== undefined) ||
      (props.domainNameOrigin !== undefined && props.certOrigin === undefined)
    ) {
      throw new Error(
        'If either of domainNameOrigin or certOrigin are set then the other must be set',
      );
    }

    if (
      (props.domainNameEdge === undefined && props.certOrigin !== undefined) ||
      (props.domainNameEdge !== undefined && props.certOrigin === undefined)
    ) {
      throw new Error(
        'If either of domainNameOrigin or certOrigin are set then the other must be set',
      );
    }

    const {
      r53Zone,
      domainNameEdge,
      domainNameOrigin,
      certOrigin,
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      rootPathPrefix,
    } = props;

    // API Gateway uses the `id` string as the gateway name without
    // any randomization... we have to make sure the name is unique-ish
    const apigatewayName = assetNameRoot
      ? `${assetNameRoot}${assetNameSuffix}`
      : `${cdk.Stack.of(this).stackName}-microapps`;

    //
    // APIGateway domain names for CloudFront and origin
    //
    this._httpApi = new apigwy.HttpApi(this, 'gwy', {
      apiName: apigatewayName,
      createDefaultStage: false,
    });
    if (removalPolicy !== undefined) {
      this._httpApi.applyRemovalPolicy(removalPolicy);
    }

    // Create the stage
    const stage = new apigwy.HttpStage(this, 'stage', {
      httpApi: this._httpApi,
      autoDeploy: true,
      // If rootPathPrefix is not defined this will be the $default stage
      stageName: rootPathPrefix,
    });

    if (domainNameEdge !== undefined && certOrigin !== undefined) {
      // Create Custom Domains for API Gateway
      const dnAppsEdge = new apigwy.DomainName(this, 'microapps-apps-edge-dn', {
        domainName: domainNameEdge,
        certificate: certOrigin,
      });
      if (removalPolicy !== undefined) {
        dnAppsEdge.applyRemovalPolicy(removalPolicy);
      }

      // Create the edge domain name mapping for the API
      new apigwy.ApiMapping(this, 'mapping', {
        api: this._httpApi,
        domainName: dnAppsEdge,
        stage,
      });
    }

    if (domainNameOrigin !== undefined && certOrigin !== undefined) {
      this._dnAppsOrigin = new apigwy.DomainName(this, 'origin-dn', {
        domainName: domainNameOrigin,
        certificate: certOrigin,
      });
      if (removalPolicy !== undefined) {
        this._dnAppsOrigin.applyRemovalPolicy(removalPolicy);
      }
    }

    // Enable access logs on API Gateway
    const apiAccessLogs = new logs.LogGroup(this, 'logs', {
      logGroupName: apigatewayName
        ? `/aws/apigwy/${apigatewayName}`
        : `/aws/apigwy/${this.httpApi.httpApiName}`,
      retention: logs.RetentionDays.TWO_WEEKS,
    });
    if (removalPolicy !== undefined) {
      apiAccessLogs.applyRemovalPolicy(removalPolicy);
    }
    // const stage = this._httpApi.defaultStage?.node.defaultChild as apigwy.CfnStage;
    (stage as unknown as apigwy.CfnStage).accessLogSettings = {
      destinationArn: apiAccessLogs.logGroupArn,
      format: JSON.stringify({
        requestId: '$context.requestId',
        userAgent: '$context.identity.userAgent',
        sourceIp: '$context.identity.sourceIp',
        requestTime: '$context.requestTime',
        requestTimeEpoch: '$context.requestTimeEpoch',
        httpMethod: '$context.httpMethod',
        path: '$context.path',
        status: '$context.status',
        protocol: '$context.protocol',
        responseLength: '$context.responseLength',
        domainName: '$context.domainName',
      }),
    };

    // Create a logging role
    // Tips: https://github.com/aws/aws-cdk/issues/11100
    const apiGwyLogRole = new iam.Role(this, 'logs-role', {
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonAPIGatewayPushToCloudWatchLogs',
        ),
      ],
      assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com'),
    });
    apiAccessLogs.grantWrite(apiGwyLogRole);

    //
    // Let API Gateway accept requests using domainNameOrigin
    // That is the origin URI that CloudFront uses for this gateway.
    // The gateway will refuse the traffic if it doesn't have the
    // domain name registered.
    //
    if (this._dnAppsOrigin !== undefined) {
      const mappingAppsApis = new apigwy.ApiMapping(this, 'api-map-origin', {
        api: this._httpApi,
        domainName: this._dnAppsOrigin,
        stage,
      });
      // 2021-12-12 - This should not be needed
      mappingAppsApis.node.addDependency(this._dnAppsOrigin);
      if (removalPolicy !== undefined) {
        mappingAppsApis.applyRemovalPolicy(removalPolicy);
      }
    }

    //
    // Create the origin name for API Gateway
    //
    if (r53Zone !== undefined && this._dnAppsOrigin) {
      const rrAppsOrigin = new r53.ARecord(this, 'origin-arecord', {
        zone: r53Zone,
        recordName: domainNameOrigin,
        target: r53.RecordTarget.fromAlias(
          new r53targets.ApiGatewayv2DomainProperties(
            this._dnAppsOrigin.regionalDomainName,
            this._dnAppsOrigin.regionalHostedZoneId,
          ),
        ),
      });
      if (removalPolicy !== undefined) {
        rrAppsOrigin.applyRemovalPolicy(removalPolicy);
      }
    }
  }
}
