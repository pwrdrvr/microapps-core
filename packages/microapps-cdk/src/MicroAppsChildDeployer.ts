import { existsSync } from 'fs';
import * as path from 'path';
import { Aws, Duration, RemovalPolicy } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as lambdaNodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

/**
 * Properties to initialize an instance of `MicroAppsChildDeployer`.
 */
export interface MicroAppsChildDeployerProps {
  /**
   * ARN of the parent Deployer Lambda Function
   */
  readonly parentDeployerLambdaARN: string;

  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: RemovalPolicy;

  /**
   * Application environment, passed as `NODE_ENV`
   * to the Router and Deployer Lambda functions
   */
  readonly appEnv: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Deployer timeout
   *
   * For larger applications this needs to be set up to 2-5 minutes for the S3 copy
   *
   * @default 2 minutes
   */
  readonly deployerTimeout?: Duration;
}

/**
 * Represents a MicroApps Child Deployer
 */
export interface IMicroAppsChildDeployer {
  /**
   * Lambda function for the Deployer
   */
  readonly deployerFunc: lambda.IFunction;
}

/**
 * Create a new MicroApps Child Deployer construct.
 */
export class MicroAppsChildDeployer extends Construct implements IMicroAppsChildDeployer {
  private _deployerFunc: lambda.Function;
  public get deployerFunc(): lambda.IFunction {
    return this._deployerFunc;
  }

  constructor(scope: Construct, id: string, props?: MicroAppsChildDeployerProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props cannot be undefined');
    }

    const {
      appEnv,
      deployerTimeout = Duration.minutes(2),
      assetNameRoot,
      assetNameSuffix,
      removalPolicy,
      parentDeployerLambdaARN,
    } = props;

    //
    // Deployer Lambda Function
    //

    // Create Deployer Lambda Function
    const deployerFuncName = assetNameRoot
      ? `${assetNameRoot}-deployer${assetNameSuffix}`
      : undefined;
    const deployerFuncProps: Omit<lambda.FunctionProps, 'handler' | 'code'> = {
      functionName: deployerFuncName,
      memorySize: 1769,
      logRetention: logs.RetentionDays.ONE_MONTH,
      runtime: lambda.Runtime.NODEJS_16_X,
      timeout: deployerTimeout,
      environment: {
        NODE_ENV: appEnv,
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
        PARENT_DEPLOYER_LAMBDA_ARN: parentDeployerLambdaARN,
      },
    };
    if (
      process.env.NODE_ENV === 'test' &&
      existsSync(path.join(__dirname, '..', '..', 'microapps-deployer', 'dist', 'index.js'))
    ) {
      // This is for local dev
      this._deployerFunc = new lambda.Function(this, 'deployer-func', {
        code: lambda.Code.fromAsset(path.join(__dirname, '..', '..', 'microapps-deployer', 'dist')),
        handler: 'index.handler',
        ...deployerFuncProps,
      });
    } else if (existsSync(path.join(__dirname, 'microapps-deployer', 'index.js'))) {
      // This is for built apps packaged with the CDK construct
      this._deployerFunc = new lambda.Function(this, 'deployer-func', {
        code: lambda.Code.fromAsset(path.join(__dirname, 'microapps-deployer')),
        handler: 'index.handler',
        ...deployerFuncProps,
      });
    } else {
      this._deployerFunc = new lambdaNodejs.NodejsFunction(this, 'deployer-func', {
        entry: path.join(__dirname, '..', '..', 'microapps-deployer', 'src', 'index.ts'),
        handler: 'handler',
        bundling: {
          minify: true,
          sourceMap: true,
        },
        ...deployerFuncProps,
      });
    }
    if (removalPolicy !== undefined) {
      this._deployerFunc.applyRemovalPolicy(removalPolicy);
    }

    // Grant full control over lambdas that indicate they are microapps
    const policyAPIManageLambdas = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ['lambda:*'],
      resources: [
        `arn:aws:lambda:${Aws.REGION}:${Aws.ACCOUNT_ID}:function:*`,
        `arn:aws:lambda:${Aws.REGION}:${Aws.ACCOUNT_ID}:function:*:*`,
      ],
      conditions: {
        StringEqualsIfExists: { 'aws:ResourceTag/microapp-managed': 'true' },
      },
    });
    this._deployerFunc.addToRolePolicy(policyAPIManageLambdas);
  }
}
