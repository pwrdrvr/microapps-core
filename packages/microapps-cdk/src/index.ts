export * from './MicroApps';
export * from './MicroAppsAPIGwy';
export * from './MicroAppsCF';
export * from './MicroAppsEdgeToOrigin';
export * from './MicroAppsS3';
export * from './MicroAppsSvcs';
export * from './MicroAppsTable';
