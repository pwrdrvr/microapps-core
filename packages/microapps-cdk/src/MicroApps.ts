import * as acm from '@aws-cdk/aws-certificatemanager';
import * as r53 from '@aws-cdk/aws-route53';
import * as cdk from '@aws-cdk/core';
import { IMicroAppsAPIGwy, MicroAppsAPIGwy } from './MicroAppsAPIGwy';
import { IMicroAppsCF, MicroAppsCF } from './MicroAppsCF';
import { IMicroAppsS3, MicroAppsS3 } from './MicroAppsS3';
import { IMicroAppsSvcs, MicroAppsSvcs } from './MicroAppsSvcs';
import { reverseDomain } from './utils/ReverseDomain';

/**
 * Props for MicroApps
 *
 */
export interface MicroAppsProps {
  /**
   * RemovalPolicy override for child resources
   *
   * Note: if set to DESTROY the S3 buckes will have `autoDeleteObjects` set to `true`
   *
   * @default - per resource default
   */
  readonly removalPolicy?: cdk.RemovalPolicy;

  /**
   * Passed to NODE_ENV of Router and Deployer Lambda functions.
   *
   * @default dev
   */
  readonly appEnv: string;

  /**
   * Optional asset name root
   *
   * @example microapps
   * @default - resource names auto assigned
   */
  readonly assetNameRoot?: string;

  /**
   * Optional asset name suffix
   *
   * @example -dev-pr-12
   * @default none
   */
  readonly assetNameSuffix?: string;

  /**
   * Route53 zone in which to create optional `domainNameEdge` record
   */
  readonly r53Zone?: r53.IHostedZone;

  /**
   * Certificate in US-East-1 for the CloudFront distribution.
   *
   */
  readonly certEdge?: acm.ICertificate;

  /**
   * Certificate in deployed region for the API Gateway.
   *
   */
  readonly certOrigin?: acm.ICertificate;

  /**
   * Use a strict S3 Bucket Policy that prevents applications
   * from reading/writing/modifying/deleting files in the S3 Bucket
   * outside of the path that is specific to their app/version.
   *
   * This setting should be used when applications are less than
   * fully trusted.
   *
   * @default false
   */
  readonly s3StrictBucketPolicy?: boolean;

  /**
   * Applies when using s3StrictBucketPolicy = true
   *
   * IAM Role or IAM User names to exclude from the DENY rules on the S3 Bucket Policy.
   *
   * Roles that are Assumed must instead have their AROA added to `s3PolicyBypassAROAs`.
   *
   * Typically any admin roles / users that need to view or manage the S3 Bucket
   * would be added to this list.
   *
   * @example ['arn:aws:iam::1234567890123:role/AdminAccess', 'arn:aws:iam::1234567890123:user/MyAdminUser']
   *
   * @see s3PolicyBypassAROAs
   */
  readonly s3PolicyBypassPrincipalARNs?: string[];

  /**
   * Applies when using s3StrictBucketPolicy = true
   *
   * AROAs of the IAM Role to exclude from the DENY rules on the S3 Bucket Policy.
   * This allows sessions that assume the IAM Role to be excluded from the
   * DENY rules on the S3 Bucket Policy.
   *
   * Typically any admin roles / users that need to view or manage the S3 Bucket
   * would be added to this list.
   *
   * Roles / users that are used directly, not assumed, can be added to `s3PolicyBypassRoleNames` instead.
   *
   * Note: This AROA must be specified to prevent this policy from locking
   * out non-root sessions that have assumed the admin role.
   *
   * The notPrincipals will only match the role name exactly and will not match
   * any session that has assumed the role since notPrincipals does not allow
   * wildcard matches and does not do wildcard matches implicitly either.
   *
   * The AROA must be used because there are only 3 Principal variables available:
   *  https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_variables.html#principaltable
   *  aws:username, aws:userid, aws:PrincipalTag
   *
   * For an assumed role, aws:username is blank, aws:userid is:
   *  [unique id AKA AROA for Role]:[session name]
   *
   * Table of unique ID prefixes such as AROA:
   *  https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html#identifiers-prefixes
   *
   * The name of the role is simply not available for an assumed role and, if it was,
   * a complicated comparison would be requierd to prevent exclusion
   * of applying the Deny Rule to roles from other accounts.
   *
   * To get the AROA with the AWS CLI:
   *   aws iam get-role --role-name ROLE-NAME
   *   aws iam get-user -â€“user-name USER-NAME
   *
   * @example [ 'AROA1234567890123' ]
   *
   * @see s3StrictBucketPolicy
   */
  readonly s3PolicyBypassAROAs?: string[];

  /**
   * Optional custom domain name for the CloudFront distribution.
   *
   * @example apps.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameEdge?: string;

  /**
   * Optional custom domain name for the API Gateway HTTPv2 API.
   *
   * @example apps-origin.pwrdrvr.com
   * @default auto-assigned
   */
  readonly domainNameOrigin?: string;

  /**
   * Path prefix on the root of the CloudFront distribution
   *
   * @example dev/
   */
  readonly rootPathPrefix?: string;

  /**
   * Create an extra Behavior (Route) for /api/ that allows
   * API routes to have a period in them.
   *
   * When false API routes with a period in the path will get routed to S3.
   *
   * When true API routes that contain /api/ in the path will get routed to API Gateway
   * even if they have a period in the path.
   *
   * @default true
   */
  readonly createAPIPathRoute?: boolean;
}

export interface IMicroApps {
  readonly cf: IMicroAppsCF;
  readonly s3: IMicroAppsS3;
  readonly svcs: IMicroAppsSvcs;
  readonly apigwy: IMicroAppsAPIGwy;
}

/**
 * Application deployment and runtime environment.
 */
export class MicroApps extends cdk.Construct implements IMicroApps {
  private _cf: MicroAppsCF;
  public get cf(): IMicroAppsCF {
    return this._cf;
  }

  private _s3: MicroAppsS3;
  public get s3(): IMicroAppsS3 {
    return this._s3;
  }

  private _apigwy: MicroAppsAPIGwy;
  public get apigwy(): IMicroAppsAPIGwy {
    return this._apigwy;
  }

  private _svcs: MicroAppsSvcs;
  public get svcs(): IMicroAppsSvcs {
    return this._svcs;
  }

  /**
   * MicroApps - Create entire stack of CloudFront, S3, API Gateway, and Lambda Functions.
   * This is the "Easy Button" construct to get started as quickly as possible.
   *
   * @param scope
   * @param id
   * @param props
   */
  constructor(scope: cdk.Construct, id: string, props?: MicroAppsProps) {
    super(scope, id);

    if (props === undefined) {
      throw new Error('props must be set');
    }

    const {
      domainNameEdge,
      domainNameOrigin,
      assetNameRoot,
      assetNameSuffix,
      r53Zone,
      certEdge,
      appEnv = 'dev',
      certOrigin,
      removalPolicy,
      s3PolicyBypassAROAs,
      s3PolicyBypassPrincipalARNs,
      s3StrictBucketPolicy,
      rootPathPrefix,
      createAPIPathRoute = true,
    } = props;

    this._s3 = new MicroAppsS3(this, 's3', {
      removalPolicy,
      bucketLogsName: domainNameEdge ? `${reverseDomain(domainNameEdge)}-logs` : undefined,
      bucketAppsName: domainNameEdge ? `${reverseDomain(domainNameEdge)}` : undefined,
      bucketAppsStagingName: domainNameEdge
        ? `${reverseDomain(domainNameEdge)}-staging`
        : undefined,
      assetNameRoot,
      assetNameSuffix,
    });
    this._apigwy = new MicroAppsAPIGwy(this, 'api', {
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      domainNameEdge,
      domainNameOrigin,
      r53Zone,
      certOrigin,
      rootPathPrefix,
    });
    this._cf = new MicroAppsCF(this, 'cft', {
      removalPolicy,
      assetNameRoot,
      assetNameSuffix,
      domainNameEdge,
      domainNameOrigin,
      httpApi: this._apigwy.httpApi,
      r53Zone,
      certEdge,
      bucketAppsOrigin: this._s3.bucketAppsOrigin,
      bucketLogs: this._s3.bucketLogs,
      rootPathPrefix,
      createAPIPathRoute,
    });
    this._svcs = new MicroAppsSvcs(this, 'svcs', {
      httpApi: this.apigwy.httpApi,
      removalPolicy,
      bucketApps: this._s3.bucketApps,
      bucketAppsOAI: this._s3.bucketAppsOAI,
      bucketAppsStaging: this._s3.bucketAppsStaging,
      assetNameRoot,
      assetNameSuffix,
      appEnv,
      s3PolicyBypassAROAs,
      s3PolicyBypassPrincipalARNs,
      s3StrictBucketPolicy,
      rootPathPrefix,
    });
  }
}
