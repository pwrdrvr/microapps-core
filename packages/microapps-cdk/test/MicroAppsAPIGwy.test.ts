/// <reference types="jest" />
import '@aws-cdk/assert/jest';
import { App, Stack } from '@aws-cdk/core';
import { MicroAppsAPIGwy } from '../src/MicroAppsAPIGwy';

describe('MicroAppsAPIGwy', () => {
  const app = new App({});

  it('works with no params', () => {
    const stack = new Stack(app, 'stack');
    // const cloudFrontDistro = new cf.Distribution(stack, 'cf-distro', {
    //   defaultBehavior: {
    //     allowedMethods: cf.AllowedMethods.ALLOW_ALL,
    //     cachePolicy: cf.CachePolicy.CACHING_DISABLED,
    //     compress: true,
    //     originRequestPolicy: cf.OriginRequestPolicy.ALL_VIEWER,
    //     origin: bucketAppsOrigin, // Just a dummy to create this distro
    //     viewerProtocolPolicy: cf.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
    //   },
    // });
    const construct = new MicroAppsAPIGwy(stack, 'construct', {});

    expect(construct).toBeDefined();
    expect(construct.dnAppsOrigin).toBeUndefined();
    expect(construct.httpApi).toBeDefined();
    expect(construct.node).toBeDefined();

    expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
  });
});
