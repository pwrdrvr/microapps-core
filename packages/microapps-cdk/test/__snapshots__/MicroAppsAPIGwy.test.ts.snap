// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MicroAppsAPIGwy works with no params 1`] = `
Object {
  "Resources": Object {
    "constructgwyE80472DA": Object {
      "Properties": Object {
        "Name": "stack-microapps",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "constructlogsBC77618C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": "/aws/apigwy/stack-microapps",
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "constructlogsrole387D81E6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "constructlogsroleDefaultPolicyD4FA927F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "constructlogsBC77618C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "constructlogsroleDefaultPolicyD4FA927F",
        "Roles": Array [
          Object {
            "Ref": "constructlogsrole387D81E6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "constructstage3B55CA4F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "constructgwyE80472DA",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
}
`;

exports[`MicroAppsAPIGwy works with params 1`] = `
Object {
  "Resources": Object {
    "cert56CA94EB": Object {
      "Properties": Object {
        "DomainName": "*.test.pwrdrvr.com",
        "DomainValidationOptions": Array [
          Object {
            "DomainName": "*.test.pwrdrvr.com",
            "ValidationDomain": "pwrdrvr.com",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "constructapimaporigin7C2B9F89": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "constructorigindnDD3F36CE",
        "constructstage3B55CA4F",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "constructgwyE80472DA",
        },
        "DomainName": "mydeploy-origin.test.pwrdrvr.com",
        "Stage": "/myprefix/",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
      "UpdateReplacePolicy": "Delete",
    },
    "constructgwyE80472DA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Name": "mydeploy-mysuffix",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
      "UpdateReplacePolicy": "Delete",
    },
    "constructlogsBC77618C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "LogGroupName": "/aws/apigwy/mydeploy-mysuffix",
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "constructlogsrole387D81E6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "constructlogsroleDefaultPolicyD4FA927F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "constructlogsBC77618C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "constructlogsroleDefaultPolicyD4FA927F",
        "Roles": Array [
          Object {
            "Ref": "constructlogsrole387D81E6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "constructmapping8BB481CC": Object {
      "DependsOn": Array [
        "constructmicroappsappsedgedn308BF2C3",
        "constructstage3B55CA4F",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Ref": "constructgwyE80472DA",
        },
        "DomainName": "mydeploy.test.pwrdrvr.com",
        "Stage": "/myprefix/",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
    },
    "constructmicroappsappsedgedn308BF2C3": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "mydeploy.test.pwrdrvr.com",
        "DomainNameConfigurations": Array [
          Object {
            "CertificateArn": Object {
              "Ref": "cert56CA94EB",
            },
            "EndpointType": "REGIONAL",
          },
        ],
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
      "UpdateReplacePolicy": "Delete",
    },
    "constructoriginarecord70039239": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::GetAtt": Array [
              "constructorigindnDD3F36CE",
              "RegionalDomainName",
            ],
          },
          "HostedZoneId": Object {
            "Fn::GetAtt": Array [
              "constructorigindnDD3F36CE",
              "RegionalHostedZoneId",
            ],
          },
        },
        "HostedZoneId": Object {
          "Ref": "zoneEB40FF1E",
        },
        "Name": "mydeploy-origin.test.pwrdrvr.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
      "UpdateReplacePolicy": "Delete",
    },
    "constructorigindnDD3F36CE": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "mydeploy-origin.test.pwrdrvr.com",
        "DomainNameConfigurations": Array [
          Object {
            "CertificateArn": Object {
              "Ref": "cert56CA94EB",
            },
            "EndpointType": "REGIONAL",
          },
        ],
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
      "UpdateReplacePolicy": "Delete",
    },
    "constructstage3B55CA4F": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "constructgwyE80472DA",
        },
        "AutoDeploy": true,
        "StageName": "/myprefix/",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "zoneEB40FF1E": Object {
      "Properties": Object {
        "Name": "test.pwrdrvr.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
  },
}
`;
